### PR TITLE
Predict and Output scaffolding

### PR DESCRIPTION
**Pull request by mtauraso**
- Predict basically works
- Most Pytorch ignite stuff has been factored out
- The results directory also basically works
- Predict is not yet outputting its results

### CODE REVIEW

File: src/fibad/config_utils.py [hunk 1]
```diff
@@ -236,16 +234,46 @@ def merge_configs(default_config: dict, user_config: dict) -> dict:
...
+        The verb name of the run.
+
+    Returns
+    -------
+    Path
+        The path created by this function
+    """
+    results_root = Path(config["general"]["results_dir"]).resolve()
+    timestamp = datetime.datetime.now().strftime("%Y%m%d-%H%m%S")
+    directory = results_root / f"{prefix}-{timestamp}"
```
Comment (drewoldag) [inline src/fibad/config_utils.py:261]: Perhaps a trivial question here, but I wanted to get your thoughts. When we were first talking about this, I envisioned a directory scheme like this: 
```
/results
   /train
   /predict
   /<verb>
```

I like what you have here more. But I'm also curious about the need to scan the directory listing and see the collated run history if that makes sense. 

By default we'll get alphabetized runs - would it be helpful to name the directories `{timestamp}-{verb}` to maintain the chronology? Maybe not worth it, since the user can just `ls -haltr` and get the directory contents sorted by creation.

Comment (mtauraso) [reply src/fibad/config_utils.py:261]: I did actually run into this during testing, where I wanted timestamp-first ordering because I needed not the current run but the previous run, and it was a little annoying with the verb being first.

I'm going to put the timestamp first, and keep things flat for now.

File: src/fibad/data_loaders/hsc_data_loader.py [hunk 2]
```diff
@@ -356,7 +356,7 @@ def _get_file(self, index: int) -> Path:
         filter = filter_names[index % self.num_filters]
         return self._file_to_path(filters[filter])
 
-    def _all_object_ids(self):
+    def ids(self):
         """Private read-only iterator over all object_ids that enforces a strict total order across
```
Comment (mtauraso) [inline src/fibad/data_loaders/hsc_data_loader.py:360]: NIT: Docstring should say Public not private

File: src/fibad/pytorch_ignite.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+from torch.utils.data import Dataset
+
+from fibad.config_utils import ConfigDict
+from fibad.data_loaders.data_loader_registry import fetch_data_loader_class
+from fibad.models.model_registry import fetch_model_class
+
+logger = logging.getLogger(__name__)
+
+
+def setup_model_and_dataset(config: ConfigDict) -> tuple:
```
Comment (mtauraso) [inline src/fibad/pytorch_ignite.py:18]: The astute reader will notice that the `data_loader` constructed in this function never leaves!

Both `train` and `predict` construct their real dataloader with a separate call to`dist_data_loader`

I think the real thing we want researchers to swap out is the data set, and I'd suggest:
* data_loader registry becomes data_set registry
* We add a `[data_set]` config section for `name=HSCDataSet` or the like
* we keep the `[data_loader]` config section to define data loader things (batch size, shuffling, num_workers)

File: src/fibad/train.py
Comment (drewoldag) [inline src/fibad/train.py:1]: I love how simple this file has become :)

File: src/fibad/predict.py [hunk 4]
```diff
@@ -1,19 +1,83 @@
...
+                RuntimeError("The path to save results for object {object_id} already exists.")
+            np.save(savepath, tensor.numpy(), allow_pickle=False)
+            write_index += 1
+
+    evaluator = create_evaluator(model, _save_batch)
+    evaluator.run(data_loader)
+    logger.info("finished evaluating...")
+
+
+def load_model_weights(config: ConfigDict, model):
```
Comment (drewoldag) [inline src/fibad/predict.py:53]: This looks good. I might look at lifting it a level so that it can be used by `train` as well when we're restarting a training run.

**Review by drewoldag — APPROVED** [review]
This looks great! Thanks for putting this together

File: src/fibad/predict.py (hunk 4 above)
Comment (mtauraso) [reply src/fibad/predict.py:53]: oohh, that is a good idea for it to end up elsewhere.
